src/compiler/checker.ts

```diff
@@ -7205,6 +7434,57 @@ module ts {
             return resolveCall(node, callSignatures, candidatesOutArray);
         }
 
+        function resolveDecorator(node: Decorator, candidatesOutArray: Signature[]): Signature {
+            let funcType = checkExpression(node.expression);
+            let apparentType = getApparentType(funcType);
+            if (apparentType === unknownType) {
+                return resolveErrorCall(node);
+            }
+            
+            let callSignatures = getSignaturesOfType(apparentType, SignatureKind.Call);
+            if (funcType === anyType || (!callSignatures.length && !(funcType.flags & TypeFlags.Union) && isTypeAssignableTo(funcType, globalFunctionType))) {
+                return resolveUntypedCall(node);
+            }
+            
+            let decoratorKind: string;
+            switch (node.parent.kind) {
+                case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.ClassExpression:
+                    decoratorKind = "class";
+                    break;
+                    
+                case SyntaxKind.Parameter:
+                    decoratorKind = "parameter";
+                    break;
+                
+                case SyntaxKind.PropertyDeclaration:
+                    decoratorKind = "property";
+                    break;
+                    
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.GetAccessor:
+                case SyntaxKind.SetAccessor:
+                    decoratorKind = "method";
+                    break;
+                
+                default:
+                    Debug.fail("Invalid decorator target.");
+                    break;
+            }
+            
+            let diagnosticChainHead = chainDiagnosticMessages(
+                /*details*/ undefined, 
+                Diagnostics.Unable_to_resolve_signature_of_0_decorator_when_called_as_an_expression, decoratorKind);
+            if (!callSignatures.length) {
+                let errorInfo = chainDiagnosticMessages(/*details*/ undefined, Diagnostics.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature);
+                errorInfo = concatenateDiagnosticMessageChains(diagnosticChainHead, errorInfo);

```

Comment:
Something feels odd about the way these diagnostic messages are being chained... Are you sure this is the best way to do it?